cni/pkg/nodeagent/informers.go

```
@@ -59,8 +63,32 @@ func setupHandlers(ctx context.Context, kubeClient kube.Client, dataplane MeshDa
 	s := &InformerHandlers{ctx: ctx, dataplane: dataplane, systemNamespace: systemNamespace}
 	s.queue = controllers.NewQueue("ambient",
 		controllers.WithGenericReconciler(s.reconcile),
-		controllers.WithMaxAttempts(5),
+		// Effectively uncapped max attempts.
+		// This is because ztunnel may connect at $any-point after we start the handler,
+		// and processing some events will continue to fail until there is a ztunnel
+		// ready to accept events fired by the handler.
+		//
+		// Note that even in this case, the failure mode for pods should
+		// always be `fail closed`, but we really have no reason to cap retry attempts
+		// caused by ztunnel not being connected, since it must (at some point).
+		//
+		// Note that an important corollary of this is that *failed pod events should be retryable*
+		// - POD ADD (ipset, iptables, ztunnel send) must be idempotent
+		// - POD DEL (ipset, iptables, ztunnel send) must be idempotent
+		//
+		// iptables case (reconciliation) is the only place where this is not always true,
+		// but that is being worked on.
+		//
+		// So max out attempt count and instead regulate with ratelimiter
+		controllers.WithMaxAttempts(math.MaxInt),
```

With infinite retry its important to make sure we actually stop trying eventually Looks like if the pod is deleted, we will stop so that is probably safe :+1:

---

we should audit how we are returning errors though...

```go
		log.Debugf("pod is now enrolled, adding to mesh")
		err := s.dataplane.AddPodToMesh(s.ctx, currentPod, podIPs, "")
		if err != nil {
			log.Warnf("AddPodToMesh returned: %v", err)
		}
```

here we never retry... not sure why? At least warrants a comment

---

BTW, other risk is that our queue uses `Event` instead of just the object name. This is fine only because we don't use `oldPod` for logic (as we have discussed). That is an issue with 5 retries, just made me look at it again

(to be clear, this LGTM - just talking out loud...)

---

> here we never retry... not sure why? At least warrants a comment

Derp, I looked at that and thought the same but forgot about commenting.

The reason that doesn't return *currently* is because we more or less decide that if we mutate the inpod rules but can't add the pod to ztunnel, it will (again) fail closed, and that pod will be detected and recycled.

That's why we already have a different error for

- captured-but-not-added-to-zt
-vs
- failed-captured

In the latter case we can retry the event.

With full iptables idempotency and reconciliation we could probably retry in the former case too, but I am still concerned about cases where `iptables` kmod support might not be available on the node, and we retry infinitely (failing on the iptables step) until someone gets wise and deletes the pod.

(there are ways around that too, but for now I'm going to skip retries non-ztunnel-add related pod errors)

---

> BTW, other risk is that our queue uses Event instead of just the object name. This is fine only because we don't use oldPod for logic (as we have discussed). That is an issue with 5 retries, just made me look at it again

yeah agreed.


One problem with only using the current pod state rather than the "historical" event in the context of *any* retries is that it creates a catch-22 for us:

1. once we annotate the pod, we will never retry it (because the newest pod generation is annotated).
2. if we inject iptables and then something fails and we do not annotate the pod, we will never remove the iptables on unenroll/remove (because we never annotated it).

To fix this I've added `ambient.istio.io/redirection=pending` - which we write on failed adds if we have already mutated the iptables, but something later fails. This lets us know we should remove the rules if the pod is removed, but also lets us know the pod isn't actually enrolled yet. It's also much easier for people to check than `istio-cni` logs. This is potentially controversial but I don't see a simpler solution and this is relatively easy to do.